File: android/src/main/java-mapboxgl/common/com/mapbox/rctmgl/components/AbstractMapFeature.java
package com.mapbox.rctmgl.components;

import android.content.Context;

import com.facebook.react.views.view.ReactViewGroup;
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView;

/**
 * Created by nickitaliano on 9/6/17.
 */

public abstract class AbstractMapFeature extends ReactViewGroup {
    public AbstractMapFeature(Context context) {
        super(context);
    }

    public abstract void addToMap(RCTMGLMapView mapView);
    public abstract void removeFromMap(RCTMGLMapView mapView);
}

File: android/src/main/java/com/rnmapbox/rnmbx/components/AbstractMapFeature.kt
package com.rnmapbox.rnmbx.components

import android.content.Context
import com.facebook.react.views.view.ReactViewGroup
import com.rnmapbox.rnmbx.components.mapview.RNMBXMapView

enum class RemovalReason {
    VIEW_REMOVAL,
    STYLE_CHANGE,
    ON_DESTROY,
    REORDER
}

abstract class AbstractMapFeature(context: Context?) : ReactViewGroup(context) {
    protected var mMapView: RNMBXMapView? = null;
    private var mWithMapViewCallbacks: Array<((RNMBXMapView) -> Unit)>? = null;

    open var requiresStyleLoad = true;

    open fun addToMap(mapView: RNMBXMapView) {
        mMapView = mapView;
        mWithMapViewCallbacks?.forEach { it(mapView) }
        mWithMapViewCallbacks = null;
    }

    // return false if you don not want to remove this feature based on reason
    open fun removeFromMap(mapView: RNMBXMapView,reason: RemovalReason) : Boolean {
        mMapView = null;
        return true;
    }

    internal fun withMapView(callback: (mapView: RNMBXMapView) -> Unit) {
        val mapView = mMapView;
        if (mapView == null) {
            val callbacks = mWithMapViewCallbacks ?: arrayOf();
            callbacks.plus(callback)
            mWithMapViewCallbacks = callbacks
        } else {
            callback(mapView)
        }
    }
}